nomos-cli/src/cmds/chat/mod.rs

```diff
@@ -0,0 +1,260 @@
+    loop {
+        terminal.draw(|f| ui::ui(f, &app)).unwrap();
+
+        if let Ok(update) = app.status_updates.try_recv() {
+            if let Status::Done = update {
+                app.message_in_flight = false;
+                app.message_status = None;
+            }
+            app.message_status = Some(update);
+            app.last_updated = Instant::now();
+        }
+
+        if let Ok(messages) = message_rx.try_recv() {
+            app.messages.extend(messages);
+        }
+
+        // Do not block rendering if there's no user input available
+        if !event::poll(Duration::from_millis(100)).unwrap() {
+            continue;
+        }
+
+        if let Event::Key(key) = event::read().unwrap() {
+            match key.code {
+                KeyCode::Enter => {
+                    if app.username.is_none() {
+                        app.username = Some(app.input.value().into());
+                    } else {
+                        // Do not allow more than one message in flight at a time for simplicity
+                        if !app.message_in_flight && !app.input.value().is_empty() {
+                            app.message_in_flight = true;
+                            app.payload_sender
+                                .send(
+                                    wire::serialize(&ChatMessage {
+                                        author: app.username.clone().unwrap(),
+                                        message: app.input.value().into(),
+                                    })
+                                    .unwrap()
+                                    .into(),
+                                )
+                                .unwrap();
+                        }
+                    }
+                    app.input.reset();
+                }
+                KeyCode::Esc => {
+                    return;
+                }
+                _ => {
+                    if !app.message_in_flight {
+                        app.input.handle_event(&Event::Key(key));
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct ChatMessage {
+    author: String,
+    message: String,
+}
+
+#[tokio::main]
+async fn check_for_messages(sender: Sender<Vec<ChatMessage>>, node: Url) {
+    let mut old_blocks = HashSet::new();
+
+    loop {
+        if let Ok(messages) = fetch_new_messages(&mut old_blocks, node.clone()).await {
+            sender.send(messages).expect("channel closed");
+        }
+        tokio::time::sleep(Duration::from_millis(100)).await;
+    }
+}
+
+async fn fetch_new_messages(
+    old_blocks: &mut HashSet<BlockId>,
+    node: Url,
+) -> Result<Vec<ChatMessage>, Box<dyn std::error::Error>> {
+    const NODE_CARNOT_INFO_PATH: &str = "carnot/info";
+
+    let mut new_messages = Vec::new();
+
+    let info = reqwest::get(node.join(NODE_CARNOT_INFO_PATH).unwrap())
```

Comment:
One way to slightly improve the polling part is to use one client for all request.
From reqwest::get docs:
```
**NOTE**: This function creates a new internal `Client` on each call,
and so should not be used if making many requests. Create a
[`Client`](https://docs.rs/reqwest/0.11.22/reqwest/struct.Client.html) instead.
```

I've seen that Da part is creating new client per reqwest too.

Reply:
Good catch!